File: backend/src/models/Template.js
import mongoose from 'mongoose'

import createSchema from './utils/createSchema'
import {Edge, Node} from './Workflow'

const Share = createSchema(
  {
    public: Boolean,
  },
  {_id: false, timestamps: false},
)

const WorkflowTemplateSchema = createSchema({
  userId: {type: String, required: true, index: true},
  name: {type: String, required: true},
  keywords: [String],
  nodes: {type: Map, of: Node, required: true},
  edges: {type: Map, of: Edge},
  root: {type: String, required: true},
  share: Share,
  backgroundImage: {type: String},
})

WorkflowTemplateSchema.methods.isPublic = function () {
  return this.share && !!this.share.public
}

const Template = mongoose.model('Template', WorkflowTemplateSchema)

export default Template

File: backend-v2/internal/models/template.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type TemplateShare struct {
	Public bool `json:"public" bson:"public"`
}

type WorkflowTemplate struct {
	TemplateID      primitive.ObjectID `json:"_id" bson:"_id"`
	UserID          string             `json:"userId" bson:"userId"`
	Name            string             `json:"name" bson:"name"`
	Keywords        []string           `json:"keywords" bson:"keywords"`
	Root            string             `json:"root" bson:"root"`
	Share           TemplateShare      `json:"share" bson:"share"`
	BackgroundImage string             `json:"backgroundImage,omitempty" bson:"backgroundImage,omitempty"`
	Nodes           map[string]Node    `json:"nodes" bson:"nodes"`
	Edges           map[string]Edge    `json:"edges" bson:"edges"`
	CreatedAt       time.Time          `json:"createdAt" bson:"createdAt"`
	UpdatedAt       time.Time          `json:"updatedAt" bson:"updatedAt"`
}

func (t WorkflowTemplate) IsPublic() bool {
	return t.Share.Public
}
